### PR TITLE
refactor: simplify routine call and result handler

### LINKED ISSUE #26: Not allowing user to pick between .call() and .send() when doing routines.endPoint()
## Description

Presently, developers have the option to choose between two methods:

```javascript
routines.SomeEndpoint.call()
// or
routines.SomeEndpoint.send()
```

This choice may cause confusion, especially for new developers. Additionally, it introduces the risk of developers unintentionally calling a read-only endpoint using .send(), which is not only inefficient (consumes gas and creates a tesseract) but also considered bad practice.

The desired behavior is to have the js-moi-sdk automatically determine the appropriate method based on the nature of the endpoint being called. 

Additionally, if a developer intends to simulate interaction with a mutating function, they should continue to use routines.someEndpoint.call() (as per the existing convention). 

However, the option of using routines.someEndpoint.send() for this purpose should be deprecated and ultimately removed - The SDK should only decide if the endpoint being called is a mutating function and internally use .send() for such interactions.

## Possible resolution
To achieve this behavior, the js-moi-sdk should emulate the pattern used by ethers.js when handling different types of functions:

In case of an error when doing a mutating call: SDK should throw an error.

How ethers handle a mutate function call -  
```javascript
const tx = await erc20Contract.transfer(toAddress, amount);
const receipt = await tx.wait();
```

How ethers handle a view function call - 
```javascript
const balance = await erc20Contract.balanceOf(targetAddress);
```

### PR DESCRIPTION
**Pull request by sarvalabs-harshrastogi**
The following pull introduces the below-mentioned enhancements and fixes:

### Changes:
- **Simplified Logic Driver**:
  - Introduced optional parameters (`fuelPrice` and `fuelLimit`) within the ix options for logic routine calls and deployment.
  - If `fuelPrice` is not explicitly provided, the system defaults to a value of 1.
  - When `fuelPrice` and `fuelLimit` are not provided, the implementation fetches fuel estimates using the `moi.FuelEstimate` method.
- **Error Handling in Contract Execution**:
  - Modified behaviour of `ix.result()` and `ix.wait()` to throw an error if the contract method execution encounters an error.
  - The thrown error object includes a `params` field. In case of failure, developers can access the receipt through `error.params.receipt`
- **Introduction of u256 and i256**:
  - Introduced new types u256 and i256.
- **Fix for Values Transfer in Estimating Fuel**:
  - Resolved an issue where estimating fuel for values transfer resulted in zero.
  - The problem was identified in the transfer_values function, which took a `Map` but when converted to `JSON` resulted in an empty object ({}). This has been addressed.
- **Clean-Up**:
  - Removed older build files cached in Git.
- **Update Example and Documentation**:
  - Updated example code and documentation to reflect the changes made.
  
---
- closes #26 #29 #27 #28

**Note**: This is a breaking change.

### Changes include
- [x] Bugfix (non-breaking change that solves an issue)
- [X] New feature (non-breaking change that adds functionality)
- [X] Breaking change (change that is not backwards-compatible and/or changes current functionality)

### Checklist
- [X] I have assigned this PR to myself
- [X] I have added at least 1 reviewer
- [X] I have tested this code
- [X] I have updated the README and other relevant documents (guides...)

### CODE REVIEW

File: packages/js-moi-providers/src/base-provider.ts [hunk 1]
```diff
@@ -1013,15 +1017,35 @@ export class BaseProvider extends AbstractProvider {
...
+                        ? ManifestCoder.decodeException(result.error)
+                        : null;
+                    
+                    if(error) {
+                        reject(new CustomError(error.data, ErrorCode.ACTION_REJECTED, {
+                            ...error,
+                            receipt
+                        }));
+                        return;
                     }
```
Comment (sarvalabs-gokul) [inline packages/js-moi-providers/src/base-provider.ts:1046]: can be written in a better way

File: packages/js-moi-providers/src/interaction.ts [hunk 2]
```diff
@@ -31,12 +31,11 @@ const serializePayload = (ixType: IxType, payload: InteractionPayload): Uint8Arr
...
-    const entries = new Map();
-
-    values.forEach((value, key) => entries.set(trimHexPrefix(key), toQuantity(value)))
-
-    return entries
+const processValues = (values: Map<string, number | bigint>): Record<string, string> => {
+    return Array.from(values).reduce((entries: Record<string, string>, [key, value]) => {
+        entries[key] = toQuantity(value);
+        return entries;
+    }, {});
```
Comment (sarvalabs-gokul) [inline packages/js-moi-providers/src/interaction.ts:38]: The return type mentioned in the comment doesn't match with the actual return value

File: packages/js-moi-signer/dist/signer.js [hunk 3]
```diff
@@ -104,6 +104,8 @@ class Signer {
      * performing validity checks.
      *
      * @param {InteractionObject} ixObject - The interaction object to prepare.
+     * @param {Options} options - The options for preparing the interaction object. (optional)
+     * @param {object} options.ignore - The fields to ignore validation during preparation. (optional)
```
Comment (sarvalabs-gokul) [inline packages/js-moi-signer/dist/signer.js:108]: why is this required?

**Review by sarvalabs-gokul — CHANGES_REQUESTED** [review]

File: packages/js-moi-logic/__tests__/logic.test.ts [hunk 4]
```diff
@@ -1,74 +1,139 @@
...
 
-    test.concurrent("should execute the erc20 routine", async() => {
-        expect(logicId).toBeDefined();
+    describe("logic driver initialized using signer", () => {
+        let logic: LogicDriver;
+
+        beforeAll(async () => {
+            if(logicId == null) {
+                throw new Error("logicId is not defined");
+            };
```
[the remark this answers is not in the document]
Comment (sarvalabs-harshrastogi) [reply packages/js-moi-logic/__tests__/logic.test.ts]: b2eb84c76d35164ed14f16d0312a9c3a86ead0dd

File: packages/js-moi-logic/__tests__/logic.test.ts [hunk 5]
```diff
@@ -1,74 +1,139 @@
...
+const HOST = "http://localhost:1600/";
+const MNEMONIC = "main story burst wonder sausage spice okay pioneer person unaware force bubble";
+const INITIAL_SUPPLY = 100000000;
+const SYMBOL = "MOI";
+const RECEIVER = "0x4cdc9a1430ca00cbaaab5dcd858236ba75e64b863d69fa799d31854e103ddf72";
+const deviationPath: string | undefined = "m/44'/6174'/0'/0/1";
+const provider = new JsonRpcProvider(HOST);
 
-        const factory = new LogicFactory(erc20, wallet);
+let signer: Signer;
```
[the remark this answers is not in the document]
Comment (sarvalabs-harshrastogi) [reply packages/js-moi-logic/__tests__/logic.test.ts]: b2eb84c76d35164ed14f16d0312a9c3a86ead0dd

File: docs/source/logic.rst [hunk 6]
```diff
@@ -277,82 +231,83 @@ Functions
...
+    const logicId = "0x0800007d70c34ed6ec...";
+    const address = "0xd13dfs...";
+
+    // Get logic driver
+    const logic = await getLogicDriver(logicId, wallet);
+
+    // Call the logic routine
+    const output = await logic.routines.GetBalance(address);
+
+    console.log(output.balance); // 1000000
```
[the remark this answers is not in the document]
Comment (sarvalabs-harshrastogi) [reply docs/source/logic.rst]: Object-based destructuring is used.

File: docs/source/logic.rst [hunk 7]
```diff
@@ -277,82 +231,83 @@ Functions
...
+
+    // Get logic driver
+    const logic = await getLogicDriver(logicId, wallet);
+
+    // Execute a mutating routine call
+    const ix = await logic.routines.Transfer(address, 1000);
+    console.log(ix.hash); //  0x010000423d3233...
+
+    const receipt = await ix.wait();
+    console.log(receipt); // { ... }
```
[the remark this answers is not in the document]
Comment (sarvalabs-harshrastogi) [reply docs/source/logic.rst:290]: Yes, developers can use `ix.result()` to obtain the returned values if the mutable function returns any data. Additional test added for the same. [Click Here](https://github.com/sarvalabs/js-moi-sdk/blob/b2eb84c76d35164ed14f16d0312a9c3a86ead0dd/packages/js-moi-logic/__tests__/logic.test.ts#L94-L103).

Included in commit: b2eb84c76d35164ed14f16d0312a9c3a86ead0dd

File: packages/js-moi-logic/__tests__/logic.test.ts [hunk 8]
```diff
@@ -1,74 +1,139 @@
...
+                expect(error.params.receipt).toBeDefined();
+            }
+        });
 
-        console.log(result)
+        it("should be able to read from persistent storage", async () => {
+            const symbol = await logic.persistentState.get("symbol");
+    
+            expect(symbol).toBe(SYMBOL);
+        });
```
[the remark this answers is not in the document]
Comment (sarvalabs-harshrastogi) [reply packages/js-moi-logic/__tests__/logic.test.ts:152]: commit b2eb84c76d35164ed14f16d0312a9c3a86ead0dd

File: packages/js-moi-logic/__tests__/utils/utils.ts [hunk 9]
```diff
@@ -1,11 +1,15 @@
...
-    const derivationPath = "m/44'/6174'/0'/0/1";
+    const derivationPath = "m/44'/6174'/7020'/0/0";
     const wallet = new Wallet(provider);
     await wallet.fromMnemonic(mnemonic, derivationPath);
 
     return wallet;
 }
+
+export const initializeProvider = (host?: string): BaseProvider => {
+    return host ? new JsonRpcProvider(host) : new VoyageProvider("babylon");
```
[the remark this answers is not in the document]
Comment (sarvalabs-harshrastogi) [reply packages/js-moi-logic/__tests__/utils/utils.ts]: removed function as this was not being used. 

commit b2eb84c76d35164ed14f16d0312a9c3a86ead0dd

File: packages/js-moi-logic/src/logic-base.ts [hunk 10]
```diff
@@ -116,29 +110,29 @@ export abstract class LogicBase extends ElementDescriptor {
...
+
+        if(option.fuelPrice != null) {
+            params.fuel_price = option.fuelPrice;
+        }
+
+        if(option.fuelLimit != null) {
+            params.fuel_limit = option.fuelLimit;
+        }
+
+        return { type, params: { ...params, ...option } }
```
[the remark this answers is not in the document]
Comment (sarvalabs-harshrastogi) [reply packages/js-moi-logic/src/logic-base.ts:138]: Certain fields, such as fuelLimit and fuelPrice, adhere to a pascal-case naming convention, but are mapped to snake-case equivalents (fuel_limit). Conversely, fields like `nonce` or `sender` do not require any conversion. The options object undergoes selective destructuring to handle these variations.

File: packages/js-moi-logic/src/logic-base.ts [hunk 11]
```diff
@@ -163,22 +163,36 @@ export abstract class LogicBase extends ElementDescriptor {
...
         const ixObject: LogicIxObject = {
             routine: routine,
             arguments: args
         } as LogicIxObject
 
-        // Define call, send, estimateFuel methods on ixObject
-        ixObject.call = (...args: any[]): Promise<InteractionCallResponse> => {
-            return this.executeRoutine(ixObject, "call", ...args) as Promise<InteractionCallResponse>
+        const DEFAULT_FUEL_PRICE = 1;
+        const DEFAULT_FUEL_LIMIT = 5000;
```
[the remark this answers is not in the document]
Comment (sarvalabs-harshrastogi) [reply packages/js-moi-logic/src/logic-base.ts]: b2eb84c76d35164ed14f16d0312a9c3a86ead0dd

File: packages/js-moi-logic/src/logic-driver.ts [hunk 12]
```diff
@@ -1,23 +1,22 @@
...
-    public readonly routines: Routines = {};
+export class LogicDriver<T extends Record<string, (...args: any) => any> = any> extends LogicDescriptor {
+    public readonly routines: Routines<T> = {} as Routines<T>;
     public readonly persistentState: PersistentState;
     public readonly ephemeralState: EphemeralState;
 
-    constructor(logicId: string, manifest: LogicManifest.Manifest, signer: Signer) {
-        super(logicId, manifest, signer)
+    constructor(logicId: string, manifest: LogicManifest.Manifest, value: Signer) {
+        super(logicId, manifest, value)
```
[the remark this answers is not in the document]
Comment (sarvalabs-harshrastogi) [reply packages/js-moi-logic/src/logic-driver.ts]: b2eb84c76d35164ed14f16d0312a9c3a86ead0dd

File: packages/js-moi-providers/src/interaction.ts (hunk 2 above)
Comment (sarvalabs-harshrastogi) [reply packages/js-moi-providers/src/interaction.ts]: b2eb84c76d35164ed14f16d0312a9c3a86ead0dd

File: packages/js-moi-signer/dist/signer.js (hunk 3 above)
Comment (sarvalabs-harshrastogi) [reply packages/js-moi-signer/dist/signer.js]: removed.

commit b2eb84c76d35164ed14f16d0312a9c3a86ead0dd

File: docs/source/logic.rst [hunk 13]
```diff
@@ -154,84 +154,38 @@ Methods
...
+    import { wallet } from "./wallet";
+
+    const factory = new LogicFactory(manifest, wallet);
+    
+    const symbol = "MOI";
+    const supply = 1000000;
+    
+    const ix = await factory.deploy("Seed!", symbol, supply);
+
+    console.log(result.logic_id); // 0x0800007d70c34ed6e...
```
[the remark this answers is not in the document]
Comment (sarvalabs-harshrastogi) [reply docs/source/logic.rst]: b2eb84c76d35164ed14f16d0312a9c3a86ead0dd

File: packages/js-moi-providers/dist/base-provider.js [hunk 14]
```diff
@@ -884,14 +888,28 @@ class BaseProvider extends abstract_provider_1.AbstractProvider {
...
-                        clearTimeout(timeoutId);
+                        return;
                     }
+                    const err = new js_moi_utils_1.CustomError(error.data, js_moi_utils_1.ErrorCode.ACTION_REJECTED, {
+                        ...error,
+                        receipt,
+                    });
+                    reject(err);
                 }
                 catch (err) {
```
Comment (sarvalabs-gokul) [inline packages/js-moi-providers/dist/base-provider.js:914]: either remove the try catch block if not required. Otherwise add a reject statement within the catch block.

**Review by sarvalabs-gokul — CHANGES_REQUESTED** [review]

File: packages/js-moi-providers/dist/base-provider.js (hunk 14 above)
Comment (sarvalabs-harshrastogi) [reply packages/js-moi-providers/dist/base-provider.js]: 8b747eb9b883d5f22023d50be0221686f6aba2d8

File: packages/js-moi-logic/__tests__/logic.test.ts [hunk 15]
```diff
@@ -1,74 +1,162 @@
...
 
-    test.concurrent("should deploy the erc20 logic", async() => {
-        expect(wallet).toBeDefined();
+const HOST = "http://localhost:1600/";
+const MNEMONIC = "main story burst wonder sausage spice okay pioneer person unaware force bubble";
+const INITIAL_SUPPLY = 100000000;
+const SYMBOL = "MOI";
+const RECEIVER = "0x4cdc9a1430ca00cbaaab5dcd858236ba75e64b863d69fa799d31854e103ddf72";
+const deviationPath: string | undefined = "m/44'/6174'/0'/0/1";
+const provider = new JsonRpcProvider(HOST);
```
Comment (sarvalabs-gokul) [inline packages/js-moi-logic/__tests__/logic.test.ts:15]: Change it to uppercase

**Review by sarvalabs-gokul — CHANGES_REQUESTED** [review]

File: packages/js-moi-logic/__tests__/logic.test.ts (hunk 15 above)
Comment (sarvalabs-harshrastogi) [reply packages/js-moi-logic/__tests__/logic.test.ts]: 448613e524e3e04e0b7e8dd0f577d0e9a6d2da32
